File: exercises/101-109_small_problems/easy_8/leading_substrings.rb
# Leading Substrings
# Write a method that returns a list of all substrings of a string that start at the beginning of the original string. The return value should be arranged in order from shortest to longest substring.

# Input: One string
# Output: An array of substrings
# Mental models: Given a string, return a list of all substrings in order of shortest to longest.

# Examples:

# substrings_at_start('abc') == ['a', 'ab', 'abc']
# substrings_at_start('a') == ['a']
# substrings_at_start('xyzzy') == ['x', 'xy', 'xyz', 'xyzz', 'xyzzy']

# Data structures
# Input is a string. Output is an array.

# Algorithm
# Create new array
# Loop String.length times
# Add a slice of string from 0 to current iteration to array

# Return array

# Code
def substrings_at_start(str)
  substrings = []

  0.upto(str.length-1) do |len|
    substrings << str[0..len]
  end

  substrings
end

p substrings_at_start('abc') == ['a', 'ab', 'abc']
p substrings_at_start('a') == ['a']
p substrings_at_start('xyzzy') == ['x', 'xy', 'xyz', 'xyzz', 'xyzzy']

File: exercises/js210_small_problems/list_processing/leadingSubstrings.js
// Leading Substrings
// Write a function that takes a string argument, and returns a list of all substrings that start from the beginning of the string, ordered from shortest to longest.

// Input: A string
// Output: An array representing all the substrings sorted from the shortest to the longest

// I guess we can use map, to split string into an array and then map it to the slice of the string

// Algorithm
// 1. Split string into an array
// 2. Map array (pass in for each element the index)
// 3. Return a slice of the string from 0 to current index + 1
//

function substringsAtStart(str) {
  return str.split('').map(function(current, index) {
    return str.slice(0, index + 1);
  });
}
// Examples:

console.log(substringsAtStart('abc'));      // ["a", "ab", "abc"]
console.log(substringsAtStart('a'));        // ["a"]
console.log(substringsAtStart('xyzzy'));    // ["x", "xy", "xyz", "xyzz", "xyzzy"]
